plugins/multifundchannel.c

```diff
@@ -0,0 +1,1643 @@
+	*/
+	sha256(&hash, destid, strlen(destid));
+	placeholder_script[0] = 0x00; /* SegWit version.  */
+	placeholder_script[1] = 0x20; /* PUSHDATA 32.  */
+	memcpy(&placeholder_script[2], &hash, sizeof(struct sha256));
+
+	return encode_scriptpubkey_to_addr(ctx, chainparams,
+					   placeholder_script);
+}
+
+/*~ Perform a dryrun `txprepare`, using placeholder addresses.
+The reason for doing a dryrun `txprepare` is:
+
+1.  It delegates handling of "all" to `txprepare`.
+2.  It ensures we have the funds available before we even
+    bother our peers with a channel opening proposal via
+    `fundchannel_start`.
+3.  It reserves the funds while we are doing (maybe lengthy)
+    network operations `fundchannel_start` and
+    `fundchannel_complete`.
+*/
+static struct command_result *
+perform_dryrun_txprepare(struct multifundchannel_command *mfc)
+{
+	struct out_req *req;
+	struct json_stream *js;
+	unsigned int i;
+
+	plugin_log(mfc->cmd->plugin, LOG_DBG,
+		   "mfc %p: dryrun txprepare.", mfc);
+
+	req = jsonrpc_request_start(mfc->cmd->plugin,
+				    mfc->cmd,
+				    "txprepare",
+				    &after_dryrun_txprepare,
+				    &mfc_forward_error,
+				    mfc);
+	js = req->js;
+
+	json_array_start(js, "outputs");
+	for (i = 0; i < tal_count(mfc->destinations); ++i) {
+		struct multifundchannel_destination *dest;
+
+		dest = &mfc->destinations[i];
+
+		dest->placeholder_addr_str
+			= create_placeholder_addr(mfc, dest->id);
+
+		json_object_start(js, NULL);
+		json_add_string(js, dest->placeholder_addr_str,
+				dest->all ? "all" :
+					    fmt_amount_sat(tmpctx,
+							   &dest->amount));
+		json_object_end(js);
+	}
+	json_array_end(js);
+
+	if (mfc->feerate_str)
+		json_add_string(js, "feerate", mfc->feerate_str);
+	json_add_u32(js, "minconf", mfc->minconf);
+	if (mfc->utxos_str)
+		json_add_jsonstr(js, "utxos", mfc->utxos_str);
+
+	return send_outreq(mfc->cmd->plugin, req);
+}
+
+/*---------------------------------------------------------------------------*/
+
+/* Function to get the results of a txprepare command.
+It loads the txid into the struct multifundchannel_command,
+and returns the parsed bitcoin transaction.
+*/
+static struct bitcoin_tx *
+get_txprepare_results(const tal_t *ctx,
+		      struct multifundchannel_command *mfc,
+		      const char *buf,
+		      const jsmntok_t *result)
+{
+	const jsmntok_t *txid_tok;
+	const jsmntok_t *tx_tok;
+	struct bitcoin_tx *tx;
+
+	/* Extract the txid.  */
+	txid_tok = json_get_member(buf, result, "txid");
+	if (!txid_tok)
+		plugin_err(mfc->cmd->plugin,
+			   "txprepare did not return 'txid': %.*s",
+			   result->end - result->start, buf + result->start);
+	mfc->txid = tal(mfc, struct bitcoin_txid);
+	if (!bitcoin_txid_from_hex(buf + txid_tok->start,
+				   txid_tok->end - txid_tok->start,
+				   mfc->txid))
+		plugin_err(mfc->cmd->plugin,
+			   "Unable to parse 'txid' from txprepare: "
+			   "%.*s",
+			   txid_tok->end - txid_tok->start,
+			   buf + txid_tok->start);
```

Comment:
Or `json_strdup` with a `tmpctx`.  But we consistently use this style in e.g. `plugins/pay.c`, `lightningd/pay.c`, and elsewhere, when reporting missing fields or incorrect types from JSON.

Reply:
Well, `json_strdup` is not equivalent since it actually does an allocation and copies. We use a mix of `end - start` + `buf + start` and the accessors because we are slowly migrating from the former to the latter, so let's not introduce new instances of the ones we are migrating away from :-)

Reply:
The advantage of `json_strdup` is that it is DRY and shorter.

     print("%.*s", json_tok_full_len(tok), json_tok_full(buf, tok));
     printf("%s", json_strdup(tmpctx, buf, tok));